db_test.go

```diff
@@ -1303,6 +1303,100 @@ func TestInitializeHeadTimestamp(t *testing.T) {
 	})
 }
 
+func TestNoEmptyBlocks(t *testing.T) {
+	db, close := openTestDB(t, nil)
+	defer close()
+	defer db.Close()
+	db.DisableCompactions()
+
+	// Test no blocks after compact with empty head.
+	testutil.Ok(t, db.compact())
+	testutil.Equals(t, 0, len(db.Blocks()))
+
+	// Test no blocks after deleting all samples from head.
+	blockRange := DefaultOptions.BlockRanges[0]
+	label := labels.FromStrings("foo", "bar")
+
+	app := db.Appender()
+	for i := int64(0); i < 6; i++ {
+		_, err := app.Add(label, i*blockRange, 0)
+		testutil.Ok(t, err)
+		_, err = app.Add(label, i*blockRange+1000, 0)
+		testutil.Ok(t, err)
+	}
+	testutil.Ok(t, app.Commit())
+
+	testutil.Ok(t, db.Delete(math.MinInt64, math.MaxInt64, labels.NewEqualMatcher("foo", "bar")))
+	uid, err := db.compactor.Write(db.dir, db.head, db.head.MinTime(), db.head.MaxTime(), nil)
+	testutil.Ok(t, err)
+	testutil.Equals(t, ulid.ULID{}, uid)
+	testutil.Ok(t, db.reload())
+	testutil.Ok(t, db.head.Truncate(db.head.MaxTime()))
```

Comment:
why do we need to truncate? the tests pass even without this.

Reply:
I tried to keep the test as if it's going ahead in time. So after compact all the samples in head are removed, so we truncate till that time (which is the max time). Test might pass, but this way looked better to me.

Reply:
Basically head is truncated till the point it is compacted.

Reply:
yep I see what you mean. So why not use `db.compact()` here as well which already includes this `Truncating` than we can also check that the head was Truncated as expected.